core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/HealthCheckedEndpointGroup.java

```diff
@@ -17,157 +17,190 @@
+    private final SessionProtocol protocol;
+    private final int port;
+    private final Backoff retryBackoff;
+    private final Function<? super ClientOptionsBuilder, ClientOptionsBuilder> clientConfigurator;
+    private final Function<? super HealthCheckerContext, ? extends AsyncCloseable> checker;
+
+    private final Map<Endpoint, DefaultHealthCheckerContext> contexts = new HashMap<>();
+    private final Set<Endpoint> healthyEndpoints = new NonBlockingHashSet<>();
+    @SuppressWarnings("FieldMayBeFinal") // Updated by `stateUpdater`
+    private volatile State state = State.UNINITIALIZED;
+
+    /**
+     * Creates a new instance. Must call {@link #init()} before using.
+     */
+    HealthCheckedEndpointGroup(EndpointGroup delegate, ClientFactory clientFactory,
+                               SessionProtocol protocol, int port, Backoff retryBackoff,
+                               Function<? super ClientOptionsBuilder, ClientOptionsBuilder> clientConfigurator,
+                               Function<? super HealthCheckerContext, ? extends AsyncCloseable> checker) {
         this.delegate = requireNonNull(delegate, "delegate");
+        this.clientFactory = requireNonNull(clientFactory, "clientFactory");
+        this.protocol = requireNonNull(protocol, "protocol");
+        this.port = port;
         this.retryBackoff = requireNonNull(retryBackoff, "retryBackoff");
+        this.clientConfigurator = requireNonNull(clientConfigurator, "clientConfigurator");
+        this.checker = requireNonNull(checker, "checker");
     }
 
     /**
-     * Update healthy servers and start to schedule healthcheck.
-     * A subclass being initialized with this constructor must call {@link #init()} before start being used.
+     * Update healthy servers and start to schedule health check. Must be called before using.
      */
-    protected void init() {
-        checkState(scheduledCheck == null, "init() must only be called once.");
+    void init() {
+        checkState(stateUpdater.compareAndSet(this, State.UNINITIALIZED, State.INITIALIZED),
+                   "init() must only be called once at the end of the construction.");
 
-        delegate.initialEndpointsFuture().join();
+        delegate.addListener(this::updateCandidates);
+        updateCandidates(delegate.initialEndpointsFuture().join());
 
-        checkAndUpdateHealthyServers().join();
-        scheduleCheckAndUpdateHealthyServers();
+        // Wait until the initial health of all endpoints are determined.
+        contexts.values().forEach(ctx -> ctx.initialCheckFuture.join());
     }
 
-    @Override
-    public void close() {
-        super.close();
-        delegate.close();
-
-        ScheduledFuture<?> scheduledCheck = this.scheduledCheck;
-        if (scheduledCheck != null) {
-            scheduledCheck.cancel(true);
-            this.scheduledCheck = null;
-        }
-    }
+    private void updateCandidates(List<Endpoint> candidates) {
+        synchronized (contexts) {
```

Comment:
Since contexts if a final field, I believe it's equivalent to synchronize on `this`, by just adding the keyword to the method or using `synchronized (this)` below.

Reply:
I think it's not a good idea to synchronized on `this` because a user can interfere with us by synchronizing on it.

Reply:
Ah didn't know of this best practice - thanks!